<?xml version='1.0'?>
<!-- <?xml version='1.0' encoding='utf-8'?> -->
<!-- <robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="panda"> -->
<robot name="panda">

  <!-- =============================================================== -->
  <!-- || This file does not contain official inertial properties   || -->
  <!-- || of panda robot. The values are from the identification    || -->
  <!-- || results published in: Identification of the Franka Emika  || -->
  <!-- || PandaRobot With Retrieval of Feasible Parameters Using    || -->
  <!-- || Penalty-Based Optimization                                || -->
  <!-- || by: Claudio Gaz, Marco Cognetti, Alexander Oliva,         || -->
  <!-- || Paolo Robuffo Giordano, Alessandro de Luca                || -->
  <!-- =============================================================== -->
    <link name="world"/>
    <link name="pedestal">
        <inertial>
            <origin xyz="0 0 0.5" rpy="0 0 0"/>
            <mass value="20"/>
            <inertia ixx="200" ixy="200" ixz="200" iyy="200" iyz="200" izz="200"/>
        </inertial>
        <visual>
            <origin xyz="1.5 0 0.5" rpy="0 0 0"/>
            <geometry>
                <box size="1 1 1"/>
            </geometry>
            <material name="Orange">
                <color rgba="1.0 0.5 0 1.0"/>
            </material>
        </visual>
        <collision>
            <origin xyz="1.5 0 0.5" rpy="0 0 0"/>
            <geometry>
                <box size="1 1 1"/>
            </geometry>
        </collision>
    </link>

    <gazebo reference="pedestal">
        <mu1>0.2</mu1>
        <mu2>0.2</mu2>
        <material>Gazebo/Orange</material>
    </gazebo>

    <joint name="world_joint" type="fixed">
        <parent link="world" />
        <child link="pedestal" />
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>
    
    <joint name="base_joint" type="fixed">
        <parent link="world"/>
        <child link="panda_link0"/>
        <origin xyz="1.5 0 1" rpy="0 0 0"/>
    </joint>

    <link name="panda_link0">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link0.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link0.stl"/>
            </geometry>
        </collision>
        <inertial>
        
            <origin rpy="0 0 0" xyz="-0.041018 -0.00014 0.049974"/>
            <mass value="0.629769"/>
            <inertia ixx="0.00315" ixy="8.2904E-07" ixz="0.00015" iyy="0.00388" iyz="8.2299E-06" izz="0.004285" />
        </inertial>
    </link>


    <link name="panda_link1">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link1.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link1.stl"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="0.003875 0.002081 -0.04762"/>
            <mass value="4.970684"/>
            <inertia ixx="0.70337" ixy="-0.00013900" ixz="0.0067720" iyy="0.70661" iyz="0.019169" izz="0.0091170" />
        </inertial>
    </link>

    <joint name="panda_joint1" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
        <origin rpy="0 0 0" xyz="0 0 0.333"/>
        <parent link="panda_link0"/>
        <child link="panda_link1"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="15" lower="-2.9671" effort="87" upper="2.9671" velocity="2.5"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link2">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link2.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link2.stl"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz=" -0.003141 -0.02872  0.003495"/>
            <mass value="0.646926"/>
            <inertia ixx="0.0079620" ixy="-3.9250e-3" ixz="1.0254e-02" iyy="2.8110e-02" iyz="7.0400e-04" izz="2.5995e-02" />
        </inertial>
    </link>

    <joint name="panda_joint2" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628"/>
        <origin rpy="-1.57 0 0" xyz="0 0 0"/>
        <parent link="panda_link1"/>
        <child link="panda_link2"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="7.5" lower="-1.8326" effort="87" upper="1.8326" velocity="2.5"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link3">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link3.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link3.stl"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="2.7518e-02 3.9252e-02 -6.6502e-02"/>
            <mass value="3.228604"/>
            <inertia ixx="3.7242e-02" ixy="-4.7610e-03" ixz="-1.1396e-02" iyy="3.6155e-02" iyz="-1.2805e-02" izz="1.0830e-02" />
        </inertial>
    </link>

    <joint name="panda_joint3" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
        <origin rpy="1.57 0 0" xyz="0 -0.316 0"/>
        <parent link="panda_link2"/>
        <child link="panda_link3"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="12.5" lower="-2.9671" effort="87" upper="2.9671" velocity="2.5"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link4">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link4.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link4.stl"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="-5.317e-02 1.04419e-01 2.7454e-02"/>
            <mass value="3.587895"/>
            <inertia ixx="2.5853e-02" ixy="7.7960e-03" ixz="-1.3320e-03" iyy="1.9552e-02" iyz="8.6410e-03" izz="2.8323e-02" />
        </inertial>
    </link>

    <joint name="panda_joint4" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="-0.0698"/>
        <origin rpy="1.57 0 0" xyz="0.0825 0 0"/>
        <parent link="panda_link3"/>
        <child link="panda_link4"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="12.5" lower="-3.1416" effort="87" upper="-0.4" velocity="2.5"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link5">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link5.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link5.stl"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="-1.1953e-02 4.1065e-02 -3.8437e-02"/>
            <mass value="1.225946"/>
            <inertia ixx="3.5549e-02" ixy="-2.1170e-03" ixz="-4.0370e-03" iyy="2.9474e-02"  iyz="2.2900e-04" izz="8.6270e-03" />
        </inertial>
    </link>

    <joint name="panda_joint5" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
        <origin rpy="-1.57 0 0" xyz="-0.0825 0.384 0"/>
        <parent link="panda_link4"/>
        <child link="panda_link5"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="15" lower="-2.9671" effort="12" upper="2.9671" velocity="3"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link6">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link6.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link6.stl"/>
            </geometry>
        </collision>
            <inertial>
            <origin rpy="0 0 0" xyz="6.0149e-02 -1.4117e-02 -1.0517e-02"/>
            <mass value="1.666555"/>
            <inertia ixx="1.9640e-03" ixy="1.0900e-04" ixz="-1.1580e-03" iyy="4.3540e-03" iyz="3.4100e-04" izz="5.4330e-03" />
        </inertial>
    </link>

    <joint name="panda_joint6" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525"/>
        <origin rpy="1.57 0 0" xyz="0 0 0"/>
        <parent link="panda_link5"/>
        <child link="panda_link6"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="20" lower="-0.0873" effort="12" upper="3.8223" velocity="3"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link7">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/visual/link7.dae"/>
            </geometry>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <geometry>
                <mesh filename="package://myur5_description/meshes/collision/link7.stl"/>
            </geometry>
        </collision>
        <inertial>
            <origin rpy="0 0 0" xyz="1.0517e-02 -4.252e-03 6.1597e-02"/>
            <mass value="7.35522e-01"/>
            <inertia ixx="1.2516e-02" ixy="-4.2800e-04" ixz="-1.1960e-03" iyy="1.0027e-02"  iyz="-7.4100e-04" izz="4.8150e-03"/>
        </inertial>
    </link>

    <joint name="panda_joint7" type="revolute">
        <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
        <origin rpy="1.57 0 0" xyz="0.088 0 0"/>
        <parent link="panda_link6"/>
        <child link="panda_link7"/>
        <axis xyz="0 0 1"/>
        <limit acceleration="20" lower="-2.9671" effort="12" upper="2.9671" velocity="3"/>
        <dynamics D="1" K="7000" mu_coulomb="0" mu_viscous="16" damping="0.003" friction="0.0"/>
    </joint>

    <link name="panda_link8" />

    <joint name="panda_joint8" type="fixed">
        <origin rpy="0 0 0" xyz="0 0 0.107"/>
        <parent link="panda_link7"/>
        <child link="panda_link8"/>
        <axis xyz="0 0 0"/>
    </joint>

    <!-- Depth Camera Link -->
    <link name="camera_link">
        <visual>
            <geometry>
            <box size="0.05 0.1 0.025"/> <!-- Just a placeholder visual -->
            </geometry>
            <material name="Gray">
            <color rgba="0.5 0.5 0.5 1.0"/>
            </material>
        </visual>
        <collision>
            <geometry>
            <box size="0.05 0.1 0.025"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0"
                    iyy="0.001" iyz="0.0"
                    izz="0.001"/>
        </inertial>
    </link>

    <!-- Fixed joint from Panda hand to camera -->
    <joint name="camera_joint" type="fixed">
        <parent link="panda_link8"/>
        <!-- Adjust xyz to place camera exactly at gripper tip -->
        <origin xyz="0 0 0" rpy="3.14 -1.57 0"/>
        <child link="camera_link"/>
    </joint>

    <gazebo reference="camera_link">
        <sensor type="camera" name="rgb_camera">
            <update_rate>30</update_rate>
            <camera>
                <horizontal_fov>1.047</horizontal_fov> <!-- ~60 degrees -->
                <image>
                    <width>640</width>
                    <height>480</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.05</near>
                    <far>5.0</far>
                </clip>
            </camera>
            <plugin name="gazebo_ros_camera" filename="libgazebo_ros_camera.so">
                <alwaysOn>true</alwaysOn>
                <updateRate>30.0</updateRate>
                <cameraName>camera</cameraName>
                <imageTopicName>image_raw</imageTopicName>
                <cameraInfoTopicName>camera_info</cameraInfoTopicName>
                <frameName>camera_link</frameName>
            </plugin>
        </sensor>
    </gazebo>




<!-- PositionJointInterface -->
<transmission name ="trans_1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint1">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_1">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </actuator>
</transmission>

<transmission name ="trans_2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint2">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_2">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        
    </actuator>
</transmission>
<transmission name ="trans_3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint3">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_3">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        
    </actuator>
</transmission>
<transmission name ="trans_4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint4">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_4">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        
    </actuator>
</transmission>

<transmission name ="trans_5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint5">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_5">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        
    </actuator>
</transmission>

<transmission name ="trans_6">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint6">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_6">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        
    </actuator>
</transmission>

<transmission name ="trans_7">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint7">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor_7">
        <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        
    </actuator>
</transmission>




<gazebo>
    <plugin name="control" filename="libgazebo_ros_control.so"/>
</gazebo>


</robot>
